### PR TITLE
fix(service_log): match template RTS QC by unit intersection instead of strict subset (#829)

### LINKED ISSUE #829: Service event templates excluded from new service event page because their RTS QC does not represent a subset of QC assigned to the unit
The title is a mouthful, but it describes the current behavior of the code. In [service_log/views.py::service_event_template_searcher](https://github.com/qatrackplus/qatrackplus/blob/aaaaf3b1635300c0e6a5f86935868b8a44860f07/qatrack/service_log/views.py#L1630), this is what happens.

In particular, this section:
```python
    # now find all templates which either have no return to service QC
    # or who's RTS QC is a subset of the QC assigned to the unit
    # Probably there is a way to express this in a single SQL query...
    sts = qs.values_list("pk", "return_to_service_test_lists", "return_to_service_cycles")
    matching_templates = []
    rts_utcs = {}
    for template_id, rts_qc_ids in groupby(sts, lambda x: x[0]):

        # get return to service test lists and test list cycles for this template
        rts_tl_ids = set(x[1] for x in rts_qc_ids if x[1] is not None)
        rts_tlc_ids = set(x[2] for x in rts_qc_ids if x[2] is not None)

        # check if there is no RTS QC for this template
        no_rts = len(rts_tl_ids) + len(rts_tlc_ids) == 0

        # check if template RTS QC is subset of QC assigned to this unit
        is_subset = (len(rts_tl_ids - unit_tls) == 0 and len(rts_tlc_ids - unit_tlcs) == 0)

        if is_subset or no_rts:
            matching_templates.append(template_id)
            # get UTC id for RTS test lists and test list cycles
            rts_utcs[template_id] = [utcs[(tl_ct.id, tl_id)] for tl_id in rts_tl_ids if tl_id is not None]
            rts_utcs[template_id] += [utcs[(tlc_ct.id, tlc_id)] for tlc_id in rts_tlc_ids if tlc_id is not None]
```

We are starting to configure service_log with templates to eventually use them. But for some reason, some of our templates refused to appear in the drop-down.

I figured out it was because the RTS QC of the templates included test lists assigned to different units. So it means that RTS QC in service event templates **MUST** only use test lists assigned to a single unit to be useful.

I wonder if we *really* have to filter out templates for which the RTS QC is not a subset of QC assigned to the currently selected service event unit.

Of course, when you get to the View All Incomplete Return To Service page, it doesn't make sense to perform lists that are not assigned to the service event unit. Maybe the filtering could happen here? Not populate the list with test lists that are not assigned to unit, just keep the ones with active assignment?

The current behavior prevents us from doing a sort of polymorphism with the service event templates. For example, say you've got two CT-scanners of different vendors, and you're changing the tube. Because they come from different vendors, the test lists are different depending on the CT (different steps/tests to do).

Service event template:
- Name: Tube change

Service event template for CT1 assignment:
- Service Event template name: Tube Change
- Unit: CT1
- RTS QC:
    - Tube change (CT1)
    - Tube change (CT2)

Service event template for CT1 assignment:
- Service Event template name: Tube Change
- Unit: CT2
- RTS QC:
    - Tube change (CT1)
    - Tube change (CT2)

So that makes 1 Service Event Template instance, 2 assignments. Same template, multiple unit assignments.

With the current behavior, we're forced to have 2 service event templates, one for each unit.

The docs do mention this in the admin guide:

> All of the fields are optional, however special attention should be payed to the Service Area and Return to Service Test Lists (Cycles) fields. A Service Event Template **may only be applied to a unit when that unit has a matching service area and the selected return to service test lists/cycles assigned to it**. If you need to create a more generic template that can be applied to a variety of unit types you may want to leave those fields blank.

So, it's mentioned, but the implications are not immediately clear. Also, with this behavior, people changing test lists assignments might break service templates without realizing it.

### PR DESCRIPTION
**Pull request by nsmela**
Closes #829

# Pull Request Template

## Summary

In service_event_template_searcher, calculate the intersection between the template's RTS test lists/cycles and the unit's active assigned QC instead of excluding templates when rts_qc - unit_qc is non-empty.

Enables polymorphic, reusable Service Event Templates across multiple machines and modalities. Update admin documentation and unit tests to verify intersection matching.

## Related Issues

Closes #829 

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature ( adds functionality, always assumed to be breaking until tested by maintainers)
- [ ] Breaking change (fix or addition that changes existing behaviour)
- [ ] Documentation only

## Target Branch

_Confirm which branch this targets (e.g. master)._

## AI Usage Disclosure

- [ ] No AI tools were used in this PR.
- [x] AI tools were used:
  - **Tools / where used:** _e.g. GitHub Copilot for boilerplate_
  - [x] I've reviewed every AI-generated line as if I wrote it myself, and confirm: no PHI shared, no license-incompatible content (Apache-2.0).

---

### Coder Tasks

- [x] Rebased / merged `develop` branch; no merge conflicts.
- [x] `uv run pre-commit run --all-files` passes (ruff lint, ruff-format, django-upgrade).
  - `uv run ruff check . --fix` to help resolve linting errors
- [x] `uv run python -Wd manage.py check` reviewed for new deprecation warnings.
- [x] `uv run pytest` full test suite passes.
- [x] `uv run python manage.py makemigrations --check --dry-run` if migrations are found, change should be marked Breaking
- [x] Docker builds succeed if touched.
- [x] Docs updated under `docs/` if behaviour, settings, or APIs changed.
- [x] `release_notes.md` updated if user-facing.
- [x] No secrets, debug code, or stray `print`/`console.log` left in.

### Review Code

- [x] Reviewer verifies the target branch.
- [x] For non-trivial changes: pulled the branch, ran it locally, and confirmed it works as described.
- [x] CI is green; test suite and `pre-commit` pass.
- [x] Reads through all changed files for anything weird or unintended.
- [x] Logic is correct; edge cases and error handling considered.
- [x] Migrations are sane and reversible where possible.
- [x] Changes are consistent with the modernization direction (no reintroduction
      of patterns being migrated away from, e.g. flake8/yapf-era conventions).
- [x] No security concerns (input validation, permissions, injection, XSS).
- [x] Comments are sufficient and not overkill.
- [x] Typos, spellcheck, professional.
- [x] Reviewer sends back review.

### Edits

- [ ] Coder fixes any issues.
- [ ] Reviewer approves.

> Maintainer review is required for breaking changes, new features, and migrations.
> For bug fixes, refactors, and docs-only changes, reviewer approval is sufficient to merge.

---

## Testing Notes

Test database created, two relevant templates made but only one is shown. Fix applied, both templates now shown.

Test data on https://github.com/nsmela/qatrackplus/tree/feat/sample-data-generator

### CODE REVIEW

File: qatrack/service_log/views.py [hunk 1]
```diff
@@ -1658,31 +1658,21 @@ def service_event_template_searcher(request):
...
-    # Probably there is a way to express this in a single SQL query...
+    # match any return to service test lists and cycles that are assigned to this unit
     sts = qs.values_list("pk", "return_to_service_test_lists", "return_to_service_cycles")
-    matching_templates = []
     rts_utcs = {}
     for template_id, rts_qc_ids in groupby(sts, lambda x: x[0]):
 
         # get return to service test lists and test list cycles for this template
         rts_tl_ids = set(x[1] for x in rts_qc_ids if x[1] is not None)
         rts_tlc_ids = set(x[2] for x in rts_qc_ids if x[2] is not None)
```
[the remark this answers is not in the document]
Comment (nsmela) [reply qatrack/service_log/views.py]: @coderabbitai fixes implemented

[the remark this answers is not in the document]
Comment (nsmela) [reply qatrack/service_log/views.py]: @coderabbitai re-review

File: qatrack/qa/static/qa/js/qa.js [hunk 2]
```diff
@@ -1036,13 +1036,13 @@ require(['jquery', 'lodash', 'moment', 'dropzone', 'autosize', 'cheekycheck', 'i
 
                     if (data.meta.work_started){
                         if (override_date)
-                            $("#id_work_started").get(0)._flatpickr.setDate(data.meta.work_started, true);
+                            $("#id_work_started").get(0)._flatpickr.setDate(moment(data.meta.work_started, [siteConfig.MOMENT_DATETIME_FMT, "DD MMM YYYY HH:mm:ss", "D MMM YYYY HH:mm:ss", "DD/MM/YYYY HH:mm:ss", "YYYY-MM-DD HH:mm:ss", moment.ISO_8601]).toDate(), true);
```
Comment (crcrewso) [inline qatrack/qa/static/qa/js/qa.js:1039]: My JS is weak, how does this and line 1045 work when date formats are specified by a site's local_settings.py.

Comment (nsmela) [reply qatrack/qa/static/qa/js/qa.js:1039]: IIRC, it's taking the datetime string produced from python and converting it into a Date javascript object the _flatpickr understands

**Review by crcrewso — APPROVED** [review]
